[24.1] The test suite no longer works offline

### Description

Up to 24.0, it was possible to run most of the test suite without Internet access, with only a handful of tests failing. However, in 24.1 a large number of tests (776, to be precise) are throwing errors when run in an environment without Internet access.

FWICS the problem stems from `pip_editable_parts` fixture that attempts to fetch and install `setuptools`:

```pytb
__________________________________________ ERROR at setup of test_get_remote_url__no_remote ___________________________________________
[gw3] linux -- Python 3.11.9 /tmp/portage/dev-python/pip-24.1/work/pip-24.1-python3_11/install/usr/bin/python3.11

pip_src = PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3/pip_src0/pip_src')
tmpdir_factory = TempPathFactory(_given_basetemp=PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3'...p/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3'), _retention_count=0, _retention_policy='all')

    @pytest.fixture(scope="session")
    def pip_editable_parts(
        pip_src: Path, tmpdir_factory: pytest.TempPathFactory
    ) -> Tuple[Path, ...]:
        pip_editable = tmpdir_factory.mktemp("pip") / "pip"
        shutil.copytree(pip_src, pip_editable, symlinks=True)
        # noxfile.py is Python 3 only
        assert compileall.compile_dir(
            pip_editable,
            quiet=1,
            rx=re.compile("noxfile.py$"),
        )
        pip_self_install_path = tmpdir_factory.mktemp("pip_self_install")
>       subprocess.check_call(
            [
                sys.executable,
                "-m",
                "pip",
                "install",
                "--target",
                pip_self_install_path,
                "-e",
                pip_editable,
            ]
        )

pip_editable = PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3/pip0/pip')
pip_self_install_path = PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3/pip_self_install0')
pip_src    = PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3/pip_src0/pip_src')
tmpdir_factory = TempPathFactory(_given_basetemp=PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3'),
                _trace=<pluggy._tracing.TagTracerSub object at 0x7fd77d708dd0>,
                _basetemp=PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3'),
                _retention_count=0,
                _retention_policy='all')

tests/conftest.py:396: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

popenargs = (['/tmp/portage/dev-python/pip-24.1/work/pip-24.1-python3_11/install/usr/bin/python3.11', '-m', 'pip', 'install', '--target', PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3/pip_self_install0'), ...],)
kwargs = {}, retcode = 1
cmd = ['/tmp/portage/dev-python/pip-24.1/work/pip-24.1-python3_11/install/usr/bin/python3.11', '-m', 'pip', 'install', '--target', PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3/pip_self_install0'), ...]

    def check_call(*popenargs, **kwargs):
        """Run command with arguments.  Wait for command to complete.  If
        the exit code was zero then return, otherwise raise
        CalledProcessError.  The CalledProcessError object will have the
        return code in the returncode attribute.
    
        The arguments are the same as for the call function.  Example:
    
        check_call(["ls", "-l"])
        """
        retcode = call(*popenargs, **kwargs)
        if retcode:
            cmd = kwargs.get("args")
            if cmd is None:
                cmd = popenargs[0]
>           raise CalledProcessError(retcode, cmd)
E           subprocess.CalledProcessError: Command '['/tmp/portage/dev-python/pip-24.1/work/pip-24.1-python3_11/install/usr/bin/python3.11', '-m', 'pip', 'install', '--target', PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3/pip_self_install0'), '-e', PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3/pip0/pip')]' returned non-zero exit status 1.

cmd        = ['/tmp/portage/dev-python/pip-24.1/work/pip-24.1-python3_11/install/usr/bin/python3.11',
 '-m',
 'pip',
 'install',
 '--target',
 PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3/pip_self_install0'),
 '-e',
 PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3/pip0/pip')]
kwargs     = {}
popenargs  = (['/tmp/portage/dev-python/pip-24.1/work/pip-24.1-python3_11/install/usr/bin/python3.11',
  '-m',
  'pip',
  'install',
  '--target',
  PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3/pip_self_install0'),
  '-e',
  PosixPath('/tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3/pip0/pip')],)
retcode    = 1

/usr/lib/python3.11/subprocess.py:413: CalledProcessError
-------------------------------------------------------- Captured stdout setup --------------------------------------------------------
Obtaining file:///tmp/portage/dev-python/pip-24.1/temp/pytest-of-portage/pytest-0/popen-gw3/pip0/pip
  Installing build dependencies: started
  Installing build dependencies: finished with status 'error'
-------------------------------------------------------- Captured stderr setup --------------------------------------------------------
  error: subprocess-exited-with-error
  
  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 1
  ╰─> [7 lines of output]
      WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f5e8ef07550>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/setuptools/
      WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f5e8ef0c110>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/setuptools/
      WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f5e8ef0cc50>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/setuptools/
      WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f5e8ef0d690>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/setuptools/
      WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f5e8ef0e350>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/setuptools/
      ERROR: Could not find a version that satisfies the requirement setuptools>=67.6.1 (from versions: none)
      ERROR: No matching distribution found for setuptools>=67.6.1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

### Expected behavior

(Most of the) tests working offline again.

### pip version

24.1

### Python version

3.11.9, 3.12.4

### OS

Gentoo Linux amd64

### How to Reproduce

1. Disable Internet access (e.g. using `unshare -n`).
2. Run `pytest`

### Output

_No response_

### Code of Conduct

- [X] I agree to follow the [PSF Code of Conduct](https://www.python.org/psf/conduct/).

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -395,6 +395,7 @@ def pip_editable_parts(
             "-m",
             "pip",
             "install",
+            "--no-build-isolation",
             "--target",
             pip_self_install_path,
             "-e",

--- a/tests/functional/test_config_settings.py
+++ b/tests/functional/test_config_settings.py
@@ -118,6 +118,7 @@ def test_config_settings_implies_pep517(
     )
     result = script.pip(
         "wheel",
+        "--no-build-isolation",
         "--config-settings",
         "FOO=Hello",
         pkg_path,

--- a/tests/functional/test_install.py
+++ b/tests/functional/test_install.py
@@ -685,7 +685,9 @@ def test_link_hash_in_dep_fails_require_hashes(
     # Build a wheel for pkga and compute its hash.
     wheelhouse = tmp_path / "wheehouse"
     wheelhouse.mkdir()
-    script.pip("wheel", "--no-deps", "-w", wheelhouse, project_path)
+    script.pip(
+        "wheel", "--no-build-isolation", "--no-deps", "-w", wheelhouse, project_path
+    )
     digest = hashlib.sha256(
         wheelhouse.joinpath("pkga-1.0-py3-none-any.whl").read_bytes()
     ).hexdigest()
@@ -903,7 +905,14 @@ def test_editable_install__local_dir_setup_requires_with_pyproject(
         "setup(name='dummy', setup_requires=['simplewheel'])\n"
     )
 
-    script.pip("install", "--find-links", shared_data.find_links, "-e", local_dir)
+    script.pip(
+        "install",
+        "--no-build-isolation",
+        "--find-links",
+        shared_data.find_links,
+        "-e",
+        local_dir,
+    )
 
 
 def test_install_pre__setup_requires_with_pyproject(

--- a/tests/functional/test_self_update.py
+++ b/tests/functional/test_self_update.py
@@ -11,12 +11,12 @@ def test_self_update_editable(script: Any, pip_src: Any) -> None:
     # Step 1. Install pip as non-editable. This is expected to succeed as
     # the existing pip in the environment is installed in editable mode, so
     # it only places a .pth file in the environment.
-    proc = script.pip("install", pip_src)
+    proc = script.pip("install", "--no-build-isolation", pip_src)
     assert proc.returncode == 0
     # Step 2. Using the pip we just installed, install pip *again*, but
     # in editable mode. This could fail, as we'll need to uninstall the running
     # pip in order to install the new copy, and uninstalling pip while it's
     # running could fail. This test is specifically to ensure that doesn't
     # happen...
-    proc = script.pip("install", "-e", pip_src)
+    proc = script.pip("install", "--no-build-isolation", "-e", pip_src)
     assert proc.returncode == 0

